File: packages/health_connector_hk_ios/ios/health_connector_hk_ios/Sources/health_connector_hk_ios/handlers/RestingHeartRateHandler.swift
import Foundation
import HealthKit

/**
 * Handler for resting heart rate data (instant quantity type)
 */
struct RestingHeartRateHandler: HealthKitQuantityHandler {
    static var supportedType: HealthDataTypeDto {
        .restingHeartRate
    }

    static var category: HealthKitDataCategory {
        .quantitySample
    }

    static func toDTO(_ sample: HKSample) throws -> HealthRecordDto {
        guard let quantitySample = sample as? HKQuantitySample else {
            throw HealthConnectorErrors.invalidArgument(
                message: "Expected HKQuantitySample, got \(type(of: sample))"
            )
        }
        guard let dto = quantitySample.toRestingHeartRateRecordDto() else {
            throw HealthConnectorErrors
                .invalidArgument(message: "Failed to convert HKQuantitySample to RestingHeartRateRecordDto")
        }
        return dto
    }

    static func toHealthKit(_ dto: HealthRecordDto) throws -> HKSample {
        guard let restingHeartRateDto = dto as? RestingHeartRateRecordDto else {
            throw HealthConnectorErrors.invalidArgument(
                message: "Expected RestingHeartRateRecordDto, got \(type(of: dto))"
            )
        }
        return try restingHeartRateDto.toHealthKit()
    }

    static func getSampleType() throws -> HKSampleType {
        try HKQuantityType.safeQuantityType(forIdentifier: .restingHeartRate)
    }

    static func extractTimestamp(_ dto: HealthRecordDto) throws -> Int64 {
        guard let restingHeartRateDto = dto as? RestingHeartRateRecordDto else {
            throw HealthConnectorErrors.invalidArgument(
                message: "Expected RestingHeartRateRecordDto, got \(type(of: dto))"
            )
        }
        return restingHeartRateDto.time
    }

    static func toStatisticsOptions(_ metric: AggregationMetricDto) throws -> HKStatisticsOptions {
        switch metric {
        case .avg:
            return .discreteAverage
        case .min:
            return .discreteMin
        case .max:
            return .discreteMax
        case .sum, .count:
            throw HealthConnectorErrors.invalidArgument(
                message: "Aggregation metric '\(metric)' not supported for resting heart rate (discrete data)",
                details: "Supported metrics: avg, min, max"
            )
        }
    }

    static func extractAggregateValue(
        from statistics: HKStatistics,
        metric: AggregationMetricDto
    ) throws -> MeasurementUnitDto {
        let quantity: HKQuantity? = switch metric {
        case .avg:
            statistics.averageQuantity()
        case .min:
            statistics.minimumQuantity()
        case .max:
            statistics.maximumQuantity()
        case .sum, .count:
            throw HealthConnectorErrors.invalidArgument(
                message: "Aggregation metric '\(metric)' not supported for resting heart rate",
                details: "Supported metrics: avg, min, max"
            )
        }

        guard let quantity else {
            throw HealthConnectorErrors.invalidArgument(
                message: "No aggregation result for metric '\(metric)'",
                details: "Statistics returned nil for restingHeartRate"
            )
        }

        let beatsPerMinute = quantity.doubleValue(for: HKUnit.count().unitDivided(by: .minute()))
        return NumericDto(unit: .numeric, value: beatsPerMinute)
    }
}

File: packages/health_connector_hc_android/android/src/main/kotlin/com/phamtunglam/health_connector_hc_android/handlers/RestingHeartRateHandler.kt
package com.phamtunglam.health_connector_hc_android.handlers

import androidx.health.connect.client.aggregate.AggregateMetric
import androidx.health.connect.client.aggregate.AggregationResult
import androidx.health.connect.client.records.Record
import androidx.health.connect.client.records.RestingHeartRateRecord
import com.phamtunglam.health_connector_hc_android.mappers.health_record_mappers.toDto
import com.phamtunglam.health_connector_hc_android.mappers.health_record_mappers.toHealthConnect
import com.phamtunglam.health_connector_hc_android.pigeon.AggregationMetricDto
import com.phamtunglam.health_connector_hc_android.pigeon.CommonAggregateRequestDto
import com.phamtunglam.health_connector_hc_android.pigeon.HealthDataTypeDto
import com.phamtunglam.health_connector_hc_android.pigeon.HealthRecordDto
import com.phamtunglam.health_connector_hc_android.pigeon.MeasurementUnitDto
import com.phamtunglam.health_connector_hc_android.pigeon.NumericDto
import com.phamtunglam.health_connector_hc_android.pigeon.NumericUnitDto
import com.phamtunglam.health_connector_hc_android.pigeon.RestingHeartRateRecordDto
import kotlin.reflect.KClass

/**
 * Handler for Resting Heart Rate health data type.
 *
 * Characteristics:
 * - Category: Instant record (single timestamp)
 * - Aggregation: Supports AVG, MIN, MAX
 * - Health Connect Type: RestingHeartRateRecord
 */
internal object RestingHeartRateHandler :
    InstantRecordHandler,
    AggregationSupportingHandler<CommonAggregateRequestDto> {
    override val supportedType: HealthDataTypeDto = HealthDataTypeDto.RESTING_HEART_RATE

    override fun toDto(record: Record): HealthRecordDto {
        require(record is RestingHeartRateRecord) {
            "Expected RestingHeartRateRecord, got ${record::class.simpleName}"
        }
        return record.toDto()
    }

    override fun toHealthConnect(dto: HealthRecordDto): Record {
        require(dto is RestingHeartRateRecordDto) {
            "Expected RestingHeartRateRecordDto, got ${dto::class.simpleName}"
        }
        return dto.toHealthConnect()
    }

    override fun getRecordClass(): KClass<out Record> = RestingHeartRateRecord::class

    override fun toAggregateMetric(request: CommonAggregateRequestDto): AggregateMetric<*> =
        when (request.aggregationMetric) {
            AggregationMetricDto.AVG -> RestingHeartRateRecord.BPM_AVG
            AggregationMetricDto.MIN -> RestingHeartRateRecord.BPM_MIN
            AggregationMetricDto.MAX -> RestingHeartRateRecord.BPM_MAX
            AggregationMetricDto.SUM, AggregationMetricDto.COUNT ->
                throw UnsupportedOperationException(
                    "Aggregation metric ${request.aggregationMetric} for RestingHeartRate. " +
                        "Supported: AVG, MIN, MAX",
                )
        }

    override fun extractAggregateValue(
        aggregationResult: AggregationResult,
        aggregateMetric: AggregateMetric<*>,
    ): MeasurementUnitDto {
        val bpm = aggregationResult[aggregateMetric] as? Long
            ?: error("Aggregation result for $aggregateMetric is null")
        return NumericDto(value = bpm.toDouble(), unit = NumericUnitDto.NUMERIC)
    }
}
